File: Assets/objects/layers/backups/NormのBackward.py
import numpy as np

def backward(dout, cache):
    """
    Layer Normalizationのbackward passを実装します。
    
    入力:
    - dout: 上流(出力層)からの勾配
    - cache: forward passから保存された値(x_normalized, gamma, x, mean, std)
    
    出力:
    - dx: 入力層への勾配 (D, )
    - dgamma: gammaパラメータへの勾配 (D, )
    - dbeta: betaパラメータへの勾配 (D, )
    """
    
    # cacheから値を取り出す
    x_normalized = cache['x_normalized']  # 形状は(D,)のベクトル
    gamma = cache['gamma']  # 形状は(D,)のベクトル
    x = cache['x']  # 形状は(D,)のベクトル
    mean = cache['mean']  # スカラー
    std = cache['std']  # スカラー
    
    # dbetaの勾配計算
    # dout: 上流(出力層)からの勾配、バッチ処理がないので形状は(D,)のベクトル
    # dbeta: betaパラメータへの勾配、形状は(D,)のベクトル。ただし、
    #        バッチ処理がない場合、全特徴量にわたる勾配を累積してスカラーとして扱われることもある
    dbeta = np.sum(dout, axis=0)

    # dgammaの勾配計算
    # dout: 上流(出力層)からの勾配、形状は(D,)のベクトル
    # x_normalized: 正規化された入力データ、形状は(D,)のベクトル
    # dgamma: gammaパラメータへの勾配、全特徴量にわたる勾配を累積したスカラー
    dgamma = np.sum(dout * x_normalized, axis=0)
    
    # dx_normalizedに関する勾配計算
    dx_normalized = dout * gamma

    # x - mean
    x_minus_mean = x - mean

    # 標準偏差の微分(ニュートン法を使用)
    epsilon=1e-6
    a = std**2 + epsilon
    x_n = 0.5  # 初期値
    for _ in range(5):  # 5回の反復で十分な精度が得られることが多い
        x_n = x_n * (1.5 - 0.5 * a * x_n**2)
    std_inv = x_n

    # dx_normalizedによる分母の逆数の勾配
    dstd_inv = np.sum(dx_normalized * x_minus_mean, axis=0)

    # 分母の勾配
    dstd = -dstd_inv / (std**2 + 1e-6)
    
    # 分散(var)の逆数をニュートン法で計算する
    epsilon = 1e-6
    a = std**2 + epsilon  # この場合、aは分散(std**2)にepsilonを加えたもの
    x_n = 0.5  # 初期値
    for _ in range(5):  # 5回の反復で十分な精度が得られることが多い
        x_n = x_n * (2 - a * x_n)
    var_inv = x_n  # var_invは分散の逆数の近似値
    
    # 分散の勾配を求めるために、分散の逆数の計算結果を用いる
    dvar = 0.5 * dstd * var_inv  # dstdは分散に対する勾配

    # x_minus_meanの2倍
    x_minus_mean_doubled = 2. * x_minus_mean

    # 分散による平均の勾配
    dmean_var_component = dvar * np.sum(x_minus_mean_doubled, axis=0) / x.shape[0]

    # 平均の勾配
    dmean = -np.sum(dx_normalized * std_inv, axis=0) + dmean_var_component

    # xに関する勾配
    dx1 = dx_normalized * std_inv
    dx2 = dmean_var_component / x.shape[0]
    dx3 = dmean / x.shape[0]

    # 最終的なdx
    dx = dx1 + dx2 + dx3

    return dx, dgamma, dbeta

File: Assets/objects/layers/backups/NormのBackward.cs
using UdonSharp;
using UnityEngine;

public class LayerNormalizationSingle : UdonSharpBehaviour
{
    public RinaNumpy rNp; // InspectorからRinaNumpyオブジェクトをアサイン

    public float[] Backward(float[] dout, float[] x_normalized, float[] gamma, float[] x, float mean, float std)
    {
        float epsilon = 1e-6f;
        
        // dbetaの勾配計算
        float dbeta = rNp.Sum_FloatArray(dout);
        
        // dgammaの勾配計算
        float[] dout_x_normalized = rNp.Multiply_FloatArray_FloatArray(dout, x_normalized);
        float dgamma = rNp.Sum_FloatArray(dout_x_normalized);
        
        // dx_normalizedに関する勾配計算
        float[] dx_normalized = rNp.Multiply_FloatArray_FloatArray(dout, gamma);

        // x - mean
        float[] x_minus_mean = rNp.Subtract_FloatArray_Float(x, mean);

        // 標準偏差の逆数
        float std_inv = 1f / (std + epsilon);

        // dx_normalizedによる分母の逆数の勾配
        float dstd_inv = rNp.Sum_FloatArray(rNp.Multiply_FloatArray_FloatArray(dx_normalized, x_minus_mean));

        // 分母の勾配
        float dstd = -dstd_inv / (std * std + epsilon);
        
        // 分散(var)の逆数
        float var_inv = 1f / (std * std + epsilon);

        // 分散の勾配
        float dvar = 0.5f * dstd * var_inv;

        // x_minus_meanの2倍
        float[] x_minus_mean_doubled = rNp.Multiply_FloatArray_Float(x_minus_mean, 2f);

        // 分散による平均の勾配
        float dmean_var_component = dvar * rNp.Sum_FloatArray(x_minus_mean_doubled) / x.Length;

        // 平均の勾配
        float dmean = -rNp.Sum_FloatArray(dx_normalized) * std_inv + dmean_var_component;

        // xに関する勾配
        float[] dx1 = rNp.Multiply_FloatArray_Float(dx_normalized, std_inv);
        float[] dx2 = rNp.OnesLike_FloatArray(x); // 全ての要素が1の配列を作成
        dx2 = rNp.Multiply_FloatArray_Float(dx2, dmean_var_component / x.Length);
        float[] dx3 = rNp.OnesLike_FloatArray(x); // 全ての要素が1の配列を作成
        dx3 = rNp.Multiply_FloatArray_Float(dx3, dmean / x.Length);

        // 最終的なdx
        float[] dx = rNp.Add_FloatArray_FloatArray(dx1, dx2);
        dx = rNp.Add_FloatArray_FloatArray(dx, dx3);

        // 出力と、キャッシュについては、最後にまとめてコメントアウトで
        // dx: 入力層への勾配
        // dgamma: gammaパラメータへの勾配
        // dbeta: betaパラメータへの勾配

        return dx; // この例ではdxのみを返す。dgammaとdbetaも必要に応じて返すように調整すること。
    }
}
